server/consumer.go

```diff
@@ -2602,7 +2602,20 @@ func (o *consumer) processNextMsgReq(_ *subscription, c *client, _ *Account, _,
 		return
 	}
 	_, msg = c.msgParts(msg)
-	o.processNextMsgRequest(reply, msg)
+
+	inlineOk := c.kind != ROUTER && c.kind != GATEWAY && c.kind != LEAF
+	if !inlineOk {
+		// Check how long we have been away from the readloop for the route or gateway or leafnode.
+		// If too long move to a separate go routine.
+		if elapsed := time.Since(c.in.start); elapsed < noBlockThresh {
```

Comment:
I am wondering if it would not be better to extend the existing `jsAPIRoutedReq` struct to contain a type of request and then have processJSAPIRoutedRequests() call o.processNextMsgRequest() or mset.getDirectRequest(). Of course, `jsAPIRoutedReq` would become a bit more bloated with fields that are not used in some API requests, etc.. but may be ok.

I am just concerned with possible "explosion" of `go o.process..` go routines and also the fact that still, when we consider that it is ok to do it inline, then the call may then block for much longer than anticipated. So with the approach I describe, changes to readCache (the start time) would not be needed and we would unconditionally process in single go routine all requests that are `!inlineOk`.

Reply:
In my high speed tests the API version was not better and worse in some respects that just simple Go routine. Also it always does that route if not client connection, I think we can do in place as long as we know we are not stalling the route or gateway.

Reply:
> we know we are not stalling the route or gateway.

But that's the thing: say last measurement was 250ms, so you decide to do it inline, but now a single message load blocks for 2s due to high contention, etc..

Reply:
Also note these two in particular are handled at the user account layer not the system later.

Reply:
Yes the first one that inlines that is bad will not allow any other behind it to be inline.. I don't think right thing to do is *always* spin it out.